proof/invariant-abstract/IpcDet_AI.thy

```diff
@@ -93,28 +93,304 @@ lemma sfi_tcb_at [wp]:
+  shows "sorted_wrt cmp xs \<Longrightarrow> filter (\<lambda>x. \<not> cmp x y) xs = dropWhile (\<lambda>x. cmp x y) xs"
+proof (induct xs)
+  case (Cons x xs)
+  have xs: "sorted_wrt cmp xs" and x: "\<forall>z\<in>set xs. cmp x z" using Cons.prems by auto
+  note eq = Cons.hyps[OF xs, symmetric]
+  show ?case
+    apply (clarsimp simp: eq filter_id_conv dest!: bspec[OF x])
+    by (drule (1) transpD[OF tr], simp)
+qed auto
+
+lemma sorted_insort_filter_eq_insort_partition:
+  assumes "transp cmp"
+  assumes "sorted_wrt cmp xs"
+  shows "insort_filter (\<lambda>x. cmp x y) x xs = insort_partition (\<lambda>x. cmp x y) x xs"
+  by (auto simp: insort_filter_def insort_partition_def
+                 sorted_filter_takeWhile[OF assms] sorted_not_filter_dropWhile[OF assms])
+
+lemma total_reflD:
+  "total {(x,y). cmp x y} \<Longrightarrow> reflp cmp \<Longrightarrow> \<not> cmp a b \<Longrightarrow> cmp b a"
+  apply (case_tac "a=b")
+   apply (fastforce dest: reflpD)
+  by (fastforce simp: total_on_def)
+
+(* FIXME: Move *)
+lemma dropWhile_dropped_P:
+  "\<lbrakk>x \<in> set queue; x \<notin> set (dropWhile P queue)\<rbrakk> \<Longrightarrow> P x"
+  by (induction queue arbitrary: x; fastforce split: if_split_asm)
+
+(* FIXME: Move *)
+lemma takeWhile_taken_P:
+  "x \<in> set (takeWhile P queue) \<Longrightarrow> P x"
+  by (induction queue arbitrary: x; fastforce split: if_split_asm)
+
+lemma sorted_insort_partition:
+  assumes tot: "total {(x,y). cmp x y}"
+  assumes tr: "transp cmp"
+  assumes re: "reflp cmp"
+  assumes sorted: "sorted_wrt cmp xs"
+  shows "sorted_wrt cmp (insort_partition (\<lambda>x. cmp x z) z xs)"
+  unfolding insort_partition_def
+  apply (clarsimp simp: sorted_wrt_append, intro conjI)
+     apply (subst sorted_filter_takeWhile[symmetric, OF tr sorted])
+     apply (rule sorted_wrt_filter, rule sorted)
+    apply (clarsimp simp: sorted_not_filter_dropWhile[symmetric, OF tr sorted])
+    apply (fastforce dest: total_reflD[OF tot re])
+   apply (subst sorted_not_filter_dropWhile[symmetric, OF tr sorted])
+   apply (rule sorted_wrt_filter, rule sorted)
+  apply (clarsimp, intro conjI)
+   apply (erule takeWhile_taken_P)
+  apply (clarsimp simp: sorted_not_filter_dropWhile[symmetric, OF tr sorted])
+  apply (drule takeWhile_taken_P)
+  apply (rule transpD[OF tr], assumption)
+  apply (fastforce dest: total_reflD[OF tot re])
+  done
+
+lemma sorted_insort_filter:
+  assumes tot: "total {(x,y). cmp x y}"
+  assumes tr: "transp cmp"
+  assumes re: "reflp cmp"
+  assumes sorted: "sorted_wrt cmp xs"
+  shows "sorted_wrt cmp (insort_filter (\<lambda>x. cmp x z) z xs)"
+  apply (subst sorted_insort_filter_eq_insort_partition[OF tr sorted])
+  by (rule sorted_insort_partition[OF tot tr re sorted])
+
+fun opt_ord_R where
+  "opt_ord_R R (Some x) (Some y) = (R x y)"
+| "opt_ord_R R x y = (y = None \<longrightarrow> x = None)"
+
+definition opt_ord where "opt_ord \<equiv> opt_ord_R (\<le>)"
+
+declare opt_ord_def[simp]
```

Comment:
this doesn't make sense ... if you have a definition where you don't even bother with a type signature, and then you immediately add that definition to `[simp]`, how is it any better than an abbreviation?

Reply:
Good point. I've made this an abbreviation